src/file/create.h

```diff
@@ -0,0 +1,169 @@
+  void *source;           // dentry for hard link, char for symlink
+} incomplete_create_t;
+
+// A map of mkdirs that have started (a kprobe) but are yet to finish
+// (the kretprobe).
+struct bpf_map_def SEC("maps/incomplete_creates") incomplete_creates = {
+  .type = BPF_MAP_TYPE_LRU_HASH,
+  .key_size = sizeof(u64),
+  .value_size = sizeof(incomplete_create_t),
+  .max_entries = 256,
+  .pinning = 0,
+  .namespace = "",
+};
+
+static __always_inline void enter_create(struct pt_regs *ctx, file_link_type_t link_type)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    incomplete_create_t event = {0};
+    event.pid_tgid = pid_tgid;
+    event.start_ktime_ns = bpf_ktime_get_ns();
+    event.link_type = link_type;
+
+    int ret = bpf_map_update_elem(&incomplete_creates, &pid_tgid, &event, BPF_ANY);
+    if (ret < 0)
+    {
+        file_message_t fm = {0};
+        fm.type = FM_WARNING;
+        fm.u.warning.pid_tgid = pid_tgid;
+        fm.u.warning.message_type.file = FM_CREATE;
+        fm.u.warning.code = W_UPDATE_MAP_ERROR;
+        fm.u.warning.info.err = ret;
+
+        push_file_message(ctx, &fm);
+    }
+}
+
+// The source parameter should be NULL if there is no source, a dentry for hard links, or a char * for symlink
+static __always_inline void store_dentry(struct pt_regs *ctx, void *path, void *dentry, void *source)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    file_message_t fm = {0};
+
+    load_event(incomplete_creates, pid_tgid, incomplete_create_t);
+    if (event.target_dentry != NULL) goto NoEvent;
+
+    event.target_dentry = dentry;
+    event.target_vfsmount = read_field_ptr(path, CRC_PATH_MNT);
+    if (event.target_vfsmount == NULL) goto EmitWarning;
+    event.source = source;
+    
+    bpf_map_update_elem(&incomplete_creates, &pid_tgid, &event, BPF_ANY);
+    return;
+
+    EventMismatch:
+    fm.type = FM_WARNING;
+    fm.u.warning.pid_tgid = pid_tgid;
+    fm.u.warning.message_type.file = FM_CREATE;
+    fm.u.warning.code = W_PID_TGID_MISMATCH;
+    fm.u.warning.info.stored_pid_tgid = event.pid_tgid;
+
+    push_file_message(ctx, &fm);
+    return;
+
+    EmitWarning:
+    push_file_warning(ctx, &fm, FM_CREATE);
+
+    NoEvent:
+    return;
+}
+
+static __always_inline void exit_create(struct pt_regs *ctx)
+{
+    u32 key = 0;
+    buf_t *buffer = (buf_t *)bpf_map_lookup_elem(&buffers, &key);
+    if (buffer == NULL) return;
+
+    cached_path_t *cached_path = (cached_path_t *)bpf_map_lookup_elem(&percpu_path, &key);
+    if (cached_path == NULL) return;
+
+    file_message_t *fm = (file_message_t *)buffer;
+    cursor_t cursor = { .buffer = buffer, .offset = &fm->u.action.buffer_len };
+    int ret = 0;
+    error_info_t info = {0};
+
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    incomplete_create_t *event = bpf_map_lookup_elem(&incomplete_creates, &pid_tgid);
+    if (event == NULL) return;
+    if (event->pid_tgid != pid_tgid) goto EventMismatch;
```

Comment:
are we doing this instead of `load_event` to save on instructions? My issue with this is that I no longer have trust in the eBPF maps that the data pointed to by `event` isn't going to be changed in a different probe while this one is running. Granted I've only ever seen this happen in the process events but I never got to the root cause that's why we began to clone the event data and *then* checking that the pid/tgid was correct.

Reply:
Previously we only accessed the event in the chunk of code that's skipped on tail calls, but we now need to fetch it on every entry.  Furthermore, we'd have to both copy everything out of the map each call, but then copy everything back in when we have to update event->target_dentry to NULL.

If you think it's really a concern, I can switch back to dealing with a copy though.

Reply:
Ah I see, and that's because we want to determine where to jump to right? I think in that case we may be overloading what this is meant to do. Maybe we want a new per cpu map to determine that?

If I knew *why* I was getting those weird mismatched events before, or if I had found a bug tracked somewhere then I would feel more strongly about it, right now it's more just a feeling of mistrust without much data on why/when the mismatch occurs.